plugins/UserCountry/Commands/AttributeHistoricalDataWithLocations.php

```diff
@@ -0,0 +1,291 @@
+
+        $output->writeln(
+            sprintf('Re-attribution for date range: %s to %s. %d visits to process with provider "%s".',
+                $from, $to, $this->amountOfVisits, $this->visitorGeolocator->getProvider()->getId())
+        );
+
+        $this->timer = new Timer();
+
+        $this->processSpecifiedLogsInChunks($output, $from, $to, $input->getOption(self::SEGMENT_LIMIT_OPTION));
+
+        $output->writeln("Completed. <comment>" . $this->timer->__toString() . "</comment>");
+
+        return 0;
+    }
+
+    protected function processSpecifiedLogsInChunks(OutputInterface $output, $from, $to, $segmentLimit)
+    {
+        $visitFieldsToSelect = array_merge(array('idvisit', 'location_ip'), array_keys(self::$logVisitFieldsToUpdate));
+
+        $lastId = 0;
+        do {
+            $logs = $this->dao->getVisitsWithDatesLimit($from, $to, $visitFieldsToSelect, $lastId, $segmentLimit);
+            if (!empty($logs)) {
+                $lastId = $logs[count($logs) - 1]['idvisit'];
+
+                $this->reattributeVisitLogs($output, $logs);
+            }
+        } while (count($logs) == $segmentLimit);
+    }
+
+    protected function reattributeVisitLogs(OutputInterface $output, $logRows)
+    {
+        foreach ($logRows as $row) {
+            if (!$this->isRowComplete($output, $row)) {
+                continue;
+            }
+
+            $location = $this->getVisitLocation($row);
+            $valuesToUpdate = $this->getValuesToUpdate($row, $location);
+
+            $this->onVisitProcessed($output);
+
+            $idVisit = $row['idvisit'];
+
+            if (empty($valuesToUpdate)) {
+                $this->writeIfVerbose($output, 'Nothing to update for idvisit = ' . $idVisit . '. Existing location info is same as geolocated.');
+            } else {
+                $this->writeIfVerbose($output, 'Updating visit with idvisit = ' . $idVisit . '.');
+
+                $this->dao->updateVisits($valuesToUpdate, $idVisit);
+                $this->dao->updateConversions($valuesToUpdate, $idVisit);
+            }
+        }
+    }
+
+    /**
+     * @param InputInterface $input
+     * @return int
+     */
+    protected function getPercentStep(InputInterface $input)
+    {
+        $percentStep = $input->getArgument(self::PERCENT_STEP_ARGUMENT);
+
+        if (!is_numeric($percentStep)) {
+            return self::PERCENT_STEP_ARGUMENT_DEFAULT;
+        }
+
+        // Percent step should be between maximum percent value and minimum percent value (1-100)
+        if ($percentStep > 99 || $percentStep < 1) {
+            return 100;
+        }
+
+        return $percentStep;
+    }
+
+    /**
+     * Returns location log values that are different than the values currently in a log row.
+     *
+     * @param array $row The visit row.
+     * @param array $location The location information.
+     * @return array The location properties to update.
+     */
+    protected function getValuesToUpdate(array $row, $location)
```

Comment:
We can also name it Gateway. In the end it shouldn't contain such code there IMO. Of course you can leave it there if you want. It would make testing easier, it would be better readable since this class does currently many things and it would allow to eg use a service/ business logic also to be used via controller if needed etc (although this will most likely not happen and is therefore not really an argument). You can write tests for this command as well probably by having protected methods in there, and in tests use another class that extends this one and makes them public. It is quite difficult to test a command this way though :( Goal must be to have tests for this.